src/client/oicp/CpoOICPClient.ts

```diff
@@ -0,0 +1,1272 @@
+
+    const oicpSession: OICPSession = {} as OICPSession;
+    oicpSession.start_datetime = transaction.timestamp;
+    oicpSession.kwh = 0;
+    oicpSession.identification = identification;
+    oicpSession.evse = oicpEvse;
+    oicpSession.currency = this.settings.currency;
+    oicpSession.status = OICPSessionStatus.PENDING;
+    oicpSession.total_cost = transaction.currentCumulatedPrice > 0 ? transaction.currentCumulatedPrice : 0;
+    oicpSession.last_updated = transaction.timestamp;
+    transaction.oicpData = {
+      session: oicpSession
+    };
+
+    // Log
+    Logging.logDebug({
+      tenantID: this.tenant.id,
+      action: ServerAction.OICP_PUSH_SESSIONS,
+      message: `Start OICP Transaction ID (ID '${transaction.id}') to Hubject`,
+      module: MODULE_NAME, method: 'startSession',
+      detailedMessages: { payload: oicpSession }
+    });
+
+    // Call Hubject
+    const response = await this.authorizeStart(transaction);
+    // Save Session ID that was created by Hubject
+    transaction.oicpData.session.id = response.SessionID;
+    transaction.oicpData.session.empPartnerSessionID = response.EMPPartnerSessionID;
+    transaction.oicpData.session.providerID = response.ProviderID;
+
+    Logging.logDebug({
+      tenantID: this.tenant.id,
+      action: ServerAction.OICP_PUSH_SESSIONS,
+      message: `Start OICP Transaction ID '${transaction.oicpData.session.id}' (ID '${transaction.id}') response received from Hubject`,
+      module: MODULE_NAME, method: 'startSession',
+      detailedMessages: { response: response }
+    });
+  }
+
+  async updateSession(transaction: Transaction): Promise<void> {
+    if (!transaction.oicpData || !transaction.oicpData.session) {
+      throw new BackendError({
+        source: transaction.chargeBoxID,
+        action: ServerAction.OICP_PUSH_SESSIONS,
+        message: 'OICP Session not started',
+        module: MODULE_NAME, method: 'updateSession',
+      });
+    }
+
+    // Hubject restriction: "Progress Notification can be sent only at interval of at least 300 seconds." (5 Minutes)
+    const date = transaction.currentTimestamp; // New Date();
+    const t1 = date.getTime();
+    const t2 = transaction.oicpData.session.last_updated.getTime();
+    const updateDuration = ((t1 - t2) / 1000);
+    console.log('update duration: ', updateDuration);
+    if (updateDuration >= 300 || transaction.oicpData.session.status === OICPSessionStatus.PENDING) {
+      transaction.oicpData.session.kwh = transaction.currentTotalConsumptionWh / 1000;
+      transaction.oicpData.session.last_updated = transaction.currentTimestamp;
+      transaction.oicpData.session.total_cost = transaction.currentCumulatedPrice > 0 ? transaction.currentCumulatedPrice : 0;
+      transaction.oicpData.session.currency = this.settings.currency;
+      // If (transaction.lastConsumption) {
+      //   transaction.oicpData.session.meterValueInBetween.push(transaction.lastConsumption.value);
+      // }
```

Comment:
If that's not part of OICP, remove dead code definitely.

Reply:
I is part of the OICP protocol but there was a problem because the value is sometimes undefined.
It is an optional field for Hubject. So, I can remove it

Reply:
It's optional but strongly recommended for realtime displaying of the pricing? If yes, add it.

Reply:
I found the bug and it works now. 
The meterValueInBetween are now included in the charging end notification and the cdr